File: mods/java/paper/src/main/java/com/alaydriem/bedrockvoicechat/paper/PaperConfigProvider.java
package com.alaydriem.bedrockvoicechat.paper;

import com.alaydriem.bedrockvoicechat.api.ConfigProvider;
import com.alaydriem.bedrockvoicechat.config.ModConfig;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.plugin.java.JavaPlugin;

/**
 * Paper-specific configuration provider using YAML files.
 */
public class PaperConfigProvider implements ConfigProvider {
    private final JavaPlugin plugin;

    public PaperConfigProvider(JavaPlugin plugin) {
        this.plugin = plugin;
    }

    @Override
    public ModConfig load() {
        FileConfiguration yamlConfig = plugin.getConfig();

        ModConfig config = new ModConfig();
        config.setBvcServer(yamlConfig.getString("bvc-server", ""));
        config.setAccessToken(yamlConfig.getString("access-token", ""));
        config.setMinimumPlayers(yamlConfig.getInt("minimum-players", 2));

        return config;
    }

    @Override
    public void save(ModConfig config) {
        FileConfiguration yamlConfig = plugin.getConfig();
        yamlConfig.set("bvc-server", config.getBvcServer());
        yamlConfig.set("access-token", config.getAccessToken());
        yamlConfig.set("minimum-players", config.getMinimumPlayers());
        plugin.saveConfig();
    }

    @Override
    public void createDefaultIfMissing() {
        plugin.saveDefaultConfig();
    }
}

File: mods/java/paper/src/main/kotlin/com/alaydriem/bedrockvoicechat/paper/PaperConfigProvider.kt
package com.alaydriem.bedrockvoicechat.paper

import com.alaydriem.bedrockvoicechat.api.ConfigProvider
import com.alaydriem.bedrockvoicechat.config.LegacyEmbeddedKeys
import com.alaydriem.bedrockvoicechat.config.ModConfig
import com.google.gson.Gson
import com.google.gson.JsonObject
import org.bukkit.plugin.java.JavaPlugin
import java.nio.file.Path

/**
 * Paper configuration provider. The YAML document is converted whole and
 * deserialized in one step, so every block the mod understands is read without
 * this class knowing any key by name.
 */
class PaperConfigProvider(private val plugin: JavaPlugin) : ConfigProvider {
    companion object {
        private val GSON = Gson()

        @JvmStatic
        fun fromJson(json: JsonObject): ModConfig {
            val config = GSON.fromJson(json, ModConfig::class.java) ?: ModConfig()
            val embedded = json.getAsJsonObject("embedded-config")
                ?: json.getAsJsonObject("embeddedConfig")
            config.legacyKeys = LegacyEmbeddedKeys.detect(embedded)
            return config
        }
    }

    override fun getConfigDir(): Path = plugin.dataFolder.toPath()

    override fun load(): ModConfig = fromJson(YamlSectionConverter.toJson(plugin.config))

    override fun createDefaultIfMissing() {
        plugin.saveDefaultConfig()
    }
}
